plugins/catalog-import/src/api/CatalogImportClient.ts

```diff
@@ -0,0 +1,192 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Octokit } from '@octokit/rest';
+import { DiscoveryApi, OAuthApi } from '@backstage/core';
+import { CatalogImportApi } from './CatalogImportApi';
+import { AnalyzeLocationResponse } from '@backstage/plugin-catalog-backend';
+import { PartialEntity } from '../util/types';
+
+export class CatalogImportClient implements CatalogImportApi {
+  private readonly discoveryApi: DiscoveryApi;
+  private readonly githubAuthApi: OAuthApi;
+
+  constructor(options: {
+    discoveryApi: DiscoveryApi;
+    githubAuthApi: OAuthApi;
+  }) {
+    this.discoveryApi = options.discoveryApi;
+    this.githubAuthApi = options.githubAuthApi;
+  }
+
+  async generateEntityDefinitions({
+    repo,
+  }: {
+    repo: string;
+  }): Promise<PartialEntity[]> {
+    const response = await fetch(
+      `${await this.discoveryApi.getBaseUrl('catalog')}/analyze-location`,
+      {
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        method: 'POST',
+        body: JSON.stringify({
+          location: { type: 'github', target: repo },
+        }),
+      },
+    ).catch(e => {
+      throw new Error(`Failed to generate entity definitions, ${e.message}`);
+    });
+    if (!response.ok) {
+      throw new Error(
+        `Failed to generate entity definitions. Received http response ${response.status}: ${response.statusText}`,
+      );
+    }
+
+    const payload = (await response.json()) as AnalyzeLocationResponse;
+    return payload.generateEntities.map(x => x.entity);
+  }
+
+  async createRepositoryLocation({
+    location,
+  }: {
+    location: string;
+  }): Promise<void> {
+    const response = await fetch(
+      `${await this.discoveryApi.getBaseUrl('catalog')}/locations`,
+      {
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        method: 'POST',
+        body: JSON.stringify({
+          type: 'github',
+          target: location,
+          presence: 'optional',
+        }),
+      },
+    );
+    if (!response.ok) {
+      throw new Error(
+        `Received http response ${response.status}: ${response.statusText}`,
+      );
+    }
+  }
+
+  async submitPrToRepo({
+    owner,
+    repo,
+    fileContent,
+  }: {
+    owner: string;
+    repo: string;
+    fileContent: string;
+  }): Promise<{ link: string; location: string }> {
+    const token = await this.githubAuthApi.getAccessToken(['repo']);
```

Comment:
We had a chat about this today, how the auth api is not very suitable at this point for this purpose. You'd probably want an api that's shaped a bit differently, where providers can be listed and auth acquired for a given one of those providers etc. Well, future work.